File: SonarApp/app/src/main/java/com/golovkobalak/sonarapp/controller/TrackingController.java
package com.golovkobalak.sonarapp.controller;

import android.util.Log;

import com.golovkobalak.sonarapp.SonarContext;
import com.golovkobalak.sonarapp.model.GeoSquare;
import com.golovkobalak.sonarapp.model.Marker;
import com.golovkobalak.sonarapp.service.TrackingService;
import com.google.gson.Gson;

import java.util.List;
import java.util.Map;

import io.javalin.Javalin;
import io.javalin.core.JavalinConfig;

public class TrackingController {

    private static final String TAG = TrackingController.class.getName();
    private static final Gson gson = new Gson();
    public static final int PORT = 8080;
    private Javalin app;
    private final TrackingService trackingService;

    public TrackingController() {
        this.trackingService = new TrackingService();
        try (Javalin appVar = Javalin.create(JavalinConfig::enableCorsForAllOrigins).start(PORT)) {
            appVar.before(ctx -> {
                Log.d(TAG, ctx.fullUrl());
                Log.d(TAG, ctx.body());
            });
            //POST http://localhost:8080/tracking/save
            appVar.post("/tracking", ctx -> {
                final String body = ctx.body();
                trackingService.saveTrackingList(body);
                ctx.status(201);
            });
            //GET http://localhost:8080/system/map/cache/dir
            appVar.get("/system/mapCacheDir", ctx -> {
                final String mapCacheDir = trackingService.getMapCacheDir();
                ctx.result(gson.toJson(mapCacheDir));
            });
            //GET http://localhost:8080/marker?north=49.960455723200724&east=36.34042262789566&south=49.955769014252176&west=36.33620619532426
            appVar.get("/marker", ctx -> {
                final Map<String, List<String>> queryParamMap = ctx.queryParamMap();
                final GeoSquare geoSquare = new GeoSquare(queryParamMap);
                final List<Marker> markers = trackingService.getMarkers(geoSquare);

                ctx.result(gson.toJson(markers));
            });
            appVar.get("/activity", ctx -> {
                ctx.result("{\"activity\":\"" + SonarContext.CURRENT_ACTIVITY + "\"}");
            });
            appVar.after(ctx -> {
                Log.d(TAG, ctx.fullUrl());
                Log.d(TAG, ctx.body());
            });
            app = appVar;
        } catch (Exception e) {
           Log.w(this.getClass().getName(), e);
        }
    }

    public void destroy() {
        app.close();
    }
}

File: SonarApp/app/src/main/java/com/golovkobalak/sonarapp/controller/TrackingController.kt
package com.golovkobalak.sonarapp.controller

import android.util.Log
import android.widget.Toast
import com.golovkobalak.sonarapp.model.GeoSquare
import com.golovkobalak.sonarapp.service.TrackingService
import com.google.gson.Gson
import io.javalin.Javalin
import io.javalin.core.JavalinConfig
import io.javalin.http.Context
import io.javalin.http.Handler

class TrackingController {
    private var currentData = ""
    private var app: Javalin? = null
    fun start() {
        val javalin = Javalin.create { obj: JavalinConfig -> obj.enableCorsForAllOrigins() }.start(PORT)
        val trackingService = TrackingService()
        javalin.before(Handler { ctx: Context ->
            Log.d(TAG, ctx.fullUrl())
            Log.d(TAG, ctx.body())
        })
        //POST http://localhost:8080/tracking/save
        javalin.post("/tracking", Handler { ctx: Context ->
            val body = ctx.body()
            trackingService.saveTrackingList(body)
            ctx.status(201)
        })
        //POST http://localhost:8080/sonar
        javalin.post("/sonar", Handler { ctx: Context ->
            val body = ctx.body()
            Log.i(this.javaClass.name, "body:" + body)
            currentData = body
            ctx.status(201)
        })
        //GET http://localhost:8080/sonar
        javalin.get("/sonar", Handler { ctx: Context ->
            ctx.result(currentData)
        })
        //GET http://localhost:8080/system/mapCacheDir
        javalin.get("/system/mapCacheDir", Handler { ctx: Context ->
            val mapCacheDir = trackingService.mapCacheDir
            ctx.result(gson.toJson(mapCacheDir))
        })
        //GET http://localhost:8080/marker?north=49.960455723200724&east=36.34042262789566&south=49.955769014252176&west=36.33620619532426
        javalin.get("/marker", Handler { ctx: Context ->
            val queryParamMap = ctx.queryParamMap()
            val geoSquare = GeoSquare(queryParamMap)
            val markers = trackingService.getMarkers(geoSquare)
            ctx.result(gson.toJson(markers))
        })
        javalin.after(Handler { ctx: Context ->
            Log.d(TAG, ctx.fullUrl())
            Log.d(TAG, ctx.body())
        })
        app = javalin
        Log.i(this.javaClass.name, "end creation")
    }

    fun destroy() {
        app?.close()
    }

    companion object {
        private val TAG = TrackingController::class.java.name
        private val gson = Gson()
        const val PORT = 8080
    }
}